### PR TITLE
V4 - BinaryKits.Zpl.Protocol

### PR DESCRIPTION
**Pull request by tinohager**
Add new project BinaryKits.Zpl.Protocol, so that the Label project could then be based on the Protocol project

```
using System;

namespace BinaryKits.Zpl.Protocol.Commands
{
    /// <summary>
    /// Command Base
    /// </summary>
    public abstract class CommandBase
    {
        /// <summary>
        /// The Command Prefix
        /// </summary>
        protected string CommandPrefix { get; private set; }

        /// <summary>
        /// Command Base
        /// </summary>
        /// <param name="commandPrefix"></param>
        public CommandBase(string commandPrefix)
        {
            this.CommandPrefix = commandPrefix;
        }

        /// <summary>
        /// Split a zpl command in data parts
        /// </summary>
        /// <param name="zplCommand"></param>
        /// <param name="dataStartIndex"></param>
        /// <returns></returns>
        protected string[] SplitCommand(string zplCommand, int dataStartIndex = 0)
        {
            var zplCommandData = zplCommand.Substring(this.CommandPrefix.Length + dataStartIndex);
            return zplCommandData.Split(',');
        }

        /// <summary>
        /// Get the Zpl Command
        /// </summary>
        /// <returns></returns>
        public abstract string ToZpl();

        /// <summary>
        /// Check the zpl command is parsable
        /// </summary>
        /// <param name="zplCommand"></param>
        /// <returns></returns>
        public bool IsCommandParsable(string zplCommand)
        {
            return zplCommand.StartsWith(this.CommandPrefix, StringComparison.OrdinalIgnoreCase);
        }

        /// <summary>
        /// Parse the Zpl Command
        /// </summary>
        /// <param name="zplCommand"></param>
        public abstract void ParseCommand(string zplCommand);
    }
}
```

### CODE REVIEW

**Comment by YipingRuan** [comment]
So `ZplFieldOrigin` in `BinaryKits.Zpl.Label` will be backed by `FieldOriginCommand` in `BinaryKits.Zpl.Protocol`, for example?

**Comment by tinohager** [comment]
**Status**

Missing Commands
- FieldBlock (^FB)
- DownloadObjects (~DY)
- GraphicField (^GF)

**Comment by tinohager** [comment]
**Status**

Missing Commands
- DownloadObjects (~DY)

**Comment by tinohager** [comment]
@YipingRuan I would have implemented the most important functions now do you have time to look through? Is actually only a very small pull request 😄

**Comment by tinohager** [comment]
284 tests are already successful.
The naming is very close to the documentation only for `StorageDevice` I have adjusted it because it was named differently in each command.

With the `DownloadObjects` command I am still not sure how to solve the switch from string to bytes on Data in a meaningful way

**Review by YipingRuan — APPROVED** [review]

File: src/BinaryKits.Zpl.Protocol/BinaryKits.Zpl.Protocol.csproj [hunk 1]
```diff
@@ -0,0 +1,42 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>net472;netstandard2.0;net5.0</TargetFrameworks>
+    <Description>This package contains the basic commands of the Zebra Programming Language. It also supports you with image conversion.</Description>
+    <Company>Binary Kits Pte. Ltd.</Company>
+    <Version>3.0.2</Version>
```
Comment (YipingRuan) [inline src/BinaryKits.Zpl.Protocol/BinaryKits.Zpl.Protocol.csproj:7]: So is this V4 or V3?

**Comment by YipingRuan** [comment]
Will this break existing Zpl.Label syntax later?

You may click merge whenever you want.

**Pull request merged**